server-utils/src/suspendable_stream.rs

```diff
@@ -0,0 +1,63 @@
+use std::time::{Duration, Instant};
+use tokio::timer::Delay;
+use std::io;
+use tokio::prelude::*;
+
+/// A wrapper type for tokio::Incoming that stops accepting new connections
+/// for a specified amount of time once `ulimit` is reached
+pub struct SuspendableStream<S> {
+	stream: S,
+	delay: Duration,
+	timeout: Option<Delay>,
+}
+
+impl<S> SuspendableStream<S> {
+	/// construct a new Suspendable stream, given tokio::Incoming
+	/// and the amount of time to pause for.
+	pub fn new(stream: S) -> Self {
+		SuspendableStream {
+			stream,
+			delay: Duration::from_millis(500),
+			timeout: None,
+		}
+	}
+}
+
+impl<S, I> Stream for SuspendableStream<S>
+	where S: Stream<Item=I, Error=io::Error>
+{
+	type Item = I;
+	type Error = ();
+
+	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
+		let mut recovered = None;
+		if let Some(mut timeout) = self.timeout.take() {
+			match timeout.poll() {
+				Ok(Async::Ready(_)) => {
+					recovered = Some(());
+				}
+				Ok(Async::NotReady) => {
+					self.timeout = Some(timeout);
+					return Ok(Async::NotReady)
+				},
+				Err(_) => unreachable!("Polling a delay shouldn't yield any errors")
```

Comment:
You should handle the error gracefully, looking into the documentation for tokio timer (https://docs.rs/tokio-timer/0.2.7/tokio_timer/struct.Error.html). You see that you can get an `at_capacity` error. As you are already in the situation, that you can not accept anymore connections, it can fairly easy be the point that you are also at the timer maximum capacity.

Reply:
Actually the only possible error is a shutdown error, taken directly from the source code.
and it occurs if the timer instance was dropped. That's not possible here.

![screenshot from 2018-10-05 09-11-25](https://user-images.githubusercontent.com/16494314/46523820-c3ba4880-c87e-11e8-96c1-e069312f005a.png)

Reply:
It is a bit tricky and, but if you look into to the code here: https://github.com/tokio-rs/tokio/blob/1879bc49ce3126af4b561bfe535cbec9f7ae5b90/tokio-timer/src/timer/entry.rs#L161
You will find functions e.g. `increment` that internally returns an `at_capacity` error. To the outside you will just get a `shutdown` error, as you already mentioned. That is somehow a design error of tokio timer.
(`register_with` is called by `register` in `Delay`.)